openhands/controller/agent_controller.py

```diff
@@ -611,27 +658,106 @@ def set_initial_state(
         else:
             self.state = state
 
-        # when restored from a previous session, the State object will have history, start_id, and end_id
-        # connect it to the event stream
-        self.state.history.set_event_stream(self.event_stream)
+            if self.state.start_id <= -1:
+                self.state.start_id = 0
 
-        # if start_id was not set in State, we're starting fresh, at the top of the stream
-        start_id = self.state.start_id
-        if start_id == -1:
-            start_id = self.event_stream.get_latest_event_id() + 1
-        else:
-            self.log(
-                'debug', f'AgentController {self.id} restoring from event {start_id}'
+            logger.debug(
+                f'AgentController {self.id} initializing history from event {self.state.start_id}'
+            )
+
+            self._init_history()
+
+    def _init_history(self):
```

Comment:
why not just serialize `history` with the rest of `state`? Due to the `close` logic?

Reply:
It's the same information we already have in the event stream. We want the event stream to be a single source of truth, that suggests we shouldn't have two.

During runtime, it's fairly easy to understand why the agent runs away with its own version of it, but I don't know, why duplicate more than that?